piksi_tools/action_logger.py

```diff
@@ -0,0 +1,242 @@
+#!/usr/bin/env python
+import piksi_tools.serial_link as sl
+from sbp.client.handler import Handler
+from sbp.logging import SBP_MSG_PRINT
+from sbp.tracking import MsgTrackingState
+from sbp.piksi import MsgMaskSatellite, SBP_MSG_RESET
+from sbp.system import SBP_MSG_HEARTBEAT
+from sbp.table import dispatch
+
+import time
+import sys
+import random
+import threading
+
+DEFAULT_POLL_INTERVAL = 60 # Seconds
+DEFAULT_MIN_SATS = 5 # min satellites to try and retain
+
+
+class LoopTimer(object):
+  """ Interval timer (emulated from from stack overflow)
+  This should get re-used on other actions during the log
+  http://stackoverflow.com/questions/12435211/python-threading-timer-repeat-function-every-n-seconds
+  """
+  def __init__(self, interval, hfunction):
+    self.interval = interval
+    self.hfunction = hfunction
+    self.thread = threading.Timer(self.interval, self.handle_function)
+
+  def handle_function(self):
+    self.hfunction()
+    self.thread = threading.Timer(self.interval, self.handle_function)
+    self.thread.start()
+
+  def start(self):
+    self.thread.start()
+
+  def cancel(self):
+    self.thread.cancel()
+
+
+
+class TestState(object):
+  """Super class for representing state-based actions during logging
+  Parameters
+  ----------
+  handler: sbp.client.handler.Handler
+      handler for SBP transfer to/from Piksi.
+  filename : string
+    File to log to.
+  """
+  def __init__(self, handler):
+    self.init_time = time.time()
+    self.handler = handler
+  def process_message(self, msg):
+    """Method to process messages from device
+    """
+    raise NotImplementedError("process_message not implemented!")
+  def action(self):
+    """Stub for communicating with device
+    """
+    pass
+
+
+class DropSatsState(TestState):
+  """
+  Instance of testState that periodically drops a random number of satellite
+  above some minimum value
+  Parameters
+  ----------
+  handler: sbp.client.handler.Handler
+      handler for SBP transfer to/from Piksi.
+  interval : int
+    number of seconds between sending mask tracking message
+  min sats : int
+    number of satellites to never go below
+  debug : bool
+    Print out extra info?
+  """
+  def __init__(self, handler, interval, min_sats, debug=False):
+    super(DropSatsState, self).__init__(handler)
+    self.min_sats = min_sats
+    self.debug = debug
+
+    # state encoding
+    self.num_tracked_sats = 0
+    self.prn_status_dict = {}
+    self.channel_status_dict = {}
+
+    # timer stuff
+    self.timer = LoopTimer(interval, self.action)
+
+  def __enter__(self):
+    self.timer.start()
+    return self
+
+  def __exit__(self, *args):
+    self.timer.cancel()
+
+  def process_message(self, msg):
+    """
+    process an SBP message into State
```

Comment:
Could be more clear, i.e. `Process SBP messages and handle resulting state transitions`